### ISSUE TITLE
Documentation for Scala 3 derives

### ISSUE THREAD

**Issue by MateuszKubuszok** [plokhotnyuk/jsoniter-scala#852]
It would be nice if documentation gave some ready-to-use solution with regards to Scala 3 and [`derives`](https://docs.scala-lang.org/scala3/reference/contextual/derivation.html) keyword.

While it's true that particular codecs might require some tuning, the user probably can come up with some defaults for their app and it might make sense to access them through derives keywords, e.g. by making a specific subclass that utilizes them.

I [tested in Scastie](https://scastie.scala-lang.org/q1jqoxMyQsG6x526PStjag) that it might be possible (I got some issues with ADTs, not sure if this is a problem with current Scala 3 macros or my configuration), and I think something like this could be put somewhere into the documentation (not necessarily my approach but some guideline), because some user could find it useful.

I believe small quality of life improvements like this would help bridge the adoption gap between jsoniter-scala and other libraries.

**Comment by plokhotnyuk** [comment]
Hi, Mateusz!

Thanks for your example! I've added it as a test and it works as expected due to recently added support of enum ADTs: https://github.com/plokhotnyuk/jsoniter-scala/commit/70a227ee35e0bfbd702d275dee3af72c9ca8dc88

I'm going to cut a release with latest improvements for Scala 3.

**Comment by MateuszKubuszok** [comment]
It works great, thanks!

**Comment by baldram** [comment]
@plokhotnyuk Hi Andriy! I tried recently to use the Type Class Derivation hoping that it "just works" with Jsoniter, but it turns out that cannot be derived since trait JsonValueCodec has no companion object.

I see, one might implement the inline `derived` manually as discussed in this thread, but maybe such default implementation could exist out-of-the-box? Still, If one needs customisation, then might use the classic `JsonCodecMaker.make` implemented explicitly in the companion object of given subject for de/serialization. But for many cases, the benefit would be that the basic use of Jsoniter is very straightforward with use of `derives`.

And this idea already works well with other libraries. An example is `zio-json`. One might do:

```scala 
final case class Foo(
    id: Long,
    bar: String
) derives JsonCodec
```
or implement it explicitly:

```scala
// a case class as above without `derives`...

object Review:
  given JsonCodec[Foo] = DeriveJsonCodec.gen[Foo]
```

So in case of Jsoniter, a basic usage example would be:
```scala
final case class Foo(
    id: Long,
    bar: String
) derives JsonValueCodec
```

What do you think?

**Comment by plokhotnyuk** [comment]
Hi Marcin,

Currently, the `derives` can be used only for a user defined extension of `JsonValueCodec`:
```scala
trait ConfiguredJsonValueCodec[A] extends JsonValueCodec[A]

object ConfiguredJsonValueCodec:
  inline def derived[A](using inline config: CodecMakerConfig = CodecMakerConfig): ConfiguredJsonValueCodec[A] = new:
    private val impl = JsonCodecMaker.make[A](config)
    export impl._
```
I'm planning to add it for the next minor version in the `jsoniter-scala-macros`  module.

Here are possible samples of usage:
```scala
{
  enum TestEnum derives ConfiguredJsonValueCodec:
    case Value1
    case Value2(string: String)

  verifySerDeser(summon[JsonValueCodec[TestEnum]], TestEnum.Value2("VVV"), """{"type":"Value2","string":"VVV"}""")
}
{
  inline given CodecMakerConfig = CodecMakerConfig.withDiscriminatorFieldName(Some("name"))

  enum TestEnum derives ConfiguredJsonValueCodec:
    case Value1
    case Value2(string: String)

  verifySerDeser(summon[JsonValueCodec[TestEnum]], TestEnum.Value2("VVV"), """{"name":"Value2","string":"VVV"}""")
}
{
  inline given CodecMakerConfig = CodecMakerConfig.withDiscriminatorFieldName(Some("hint")).withFieldNameMapper {
    case "string" => "str"
  }

  enum TestEnum derives ConfiguredJsonValueCodec:
    case Value1
    case Value2(string: String)

  verifySerDeser(summon[JsonValueCodec[TestEnum]], TestEnum.Value2("VVV"), """{"hint":"Value2","str":"VVV"}""")
}
```

Unfortunately `JsonCodec` is to generic and cannot be easily implemented for all classes.

Also, please use `derives` wisely  (only for top-level classes and for big/nested commonly used classes).

The drawback of `derives` is fixing the model to use only one configuration for the data model, while using `given` allow to define different codecs in different scopes for the same data model.

**Comment by plokhotnyuk** [comment]
@MateuszKubuszok @baldram `ConfiguredJsonValueCodec` is available for Scala 3 macros since v2.19.0
